crates/red_knot_python_semantic/resources/mdtest/protocols.md

```diff
@@ -1,9 +1,1015 @@
+```
+
+And as a corollary, `type[MyProtocol]` can also be called:
+
+```py
+def f(x: type[MyProtocol]):
+    reveal_type(x())  # revealed: MyProtocol
+```
+
+## Members of a protocol
+
+A protocol defines an interface through its *members*: if a protocol `Foo` has members `X` and `Y`,
+a type `Bar` can only be a subtype of `Foo` if inhabitants of `Bar` also have attributes `X` and `Y`.
+
+A protocol class defines its members through declarations in the class body.
+The members of a protocol can be introspected using the function `typing.get_protocol_members`,
+which is backported via `typing_extensions`.
+
+```py
+from typing_extensions import Protocol, get_protocol_members
+
+# TODO: should not error
+class Foo(Protocol):  # error: [invalid-base]
+    x: int
+
+    @property
+    def y(self) -> str:
+        return "y"
+
+    @property
+    def z(self) -> int:
+        return 42
+
+    @z.setter
+    def z(self, z: int) -> None: ...
+    def method_member(self) -> bytes:
+        return b"foo"
+
+# TODO: at runtime, `get_protocol_members` returns a `frozenset`,
+# but for now we might pretend it returns a `tuple`, as we support heterogeneous `tuple` types
+# but not yet generic `frozenset`s
+#
+# So this should either be
+#
+# `tuple[Literal["x"], Literal["y"], Literal["z"], Literal["method_member"]]`
+#
+# `frozenset[Literal["x", "y", "z", "method_member"]]`
+reveal_type(get_protocol_members(Foo))  # revealed: @Todo(generics)
+```
+
+Calling `get_protocol_members` on a non-protocol class raises an error at runtime:
+
+```py
+class NotAProtocol: ...
+
+# TODO: should emit `[invalid-protocol]` error, should reveal `Unknown`
+reveal_type(get_protocol_members(NotAProtocol))  # revealed: @Todo(generics)
+```
+
+Certain special attributes and methods are not considered protocol members at runtime, and should not be considered
+protocol members by type checkers either:
+
+```py
+# TODO: should not error
+class Lumberjack(Protocol):  # error: [invalid-base]
+    __slots__ = ()
+    __match_args__ = ()
+    x: int
+
+    def __new__(cls, x: int) -> "Lumberjack":
+        return object.__new__(cls)
+
+    def __init__(self, x: int) -> None:
+        self.x = x
+
+# TODO: `tuple[Literal["x"]]` or `frozenset[Literal["x"]]`
+reveal_type(get_protocol_members(Lumberjack))  # revealed: @Todo(generics)
+```
+
+## Subtyping of protocols with attribute members
+
+In the following example, the protocol class `HasX` defines an interface such that any other fully
+static type can be said to be a subtype of `HasX` if all inhabitants of that other type have a
+mutable `x` attribute of type `int`:
+
+```py
+from typing import Protocol
+from knot_extensions import static_assert, is_assignable_to, is_subtype_of
+
+class HasX(Protocol):
+    x: int
+
+class Foo:
+    x: int
+
+static_assert(is_subtype_of(Foo, HasX))  # error: [static-assert-error]
+static_assert(is_assignable_to(Foo, HasX))  # error: [static-assert-error]
```

Comment:
I think there should be TODO comments here?

There are enough cases of this below that maybe it's intentional to assume any `static-assert-error` is implicitly a TODO? I don't think I like this, though; too subtle and confusing, when there are other diagnostics that are correct and not TODOs.

Reply:
Hah, yes, sorry about that. I got lazy and thought I might be able to get away without explicit TODO comments for `static-assert-error`s in this test suite, since there's no reason you'd have a `static-assert-error` in this Markdown file if it wasn't something that you wanted to pass eventually.

I'll add the TODOs :-)